components/style/gecko_conversions.rs

```diff
@@ -79,3 +79,184 @@ impl From<nsStyleCoord_CalcValue> for LengthOrPercentage {
         }
     }
 }
+
+pub mod basic_shape {
+    use euclid::size::Size2D;
+    use gecko_bindings::structs::StyleClipPathGeometryBox;
+    use gecko_bindings::structs::{StyleBasicShape, StyleBasicShapeType, StyleFillRule};
+    use gecko_bindings::structs::{nsStyleCoord, nsStyleCorners, nsStyleImageLayers_Position};
+    use gecko_bindings::sugar::ns_style_coord::{CoordDataMut, CoordDataValue};
+    use gecko_values::GeckoStyleCoordConvertible;
+    use std::borrow::Borrow;
+    use values::computed::basic_shape::*;
+    use values::computed::position::Position;
+    use values::computed::{BorderRadiusSize, LengthOrPercentage};
+
+    // using Borrow so that we can have a non-moving .into()
+    impl<T: Borrow<StyleBasicShape>> From<T> for BasicShape {
+        fn from(other: T) -> Self {
+            let other = other.borrow();
```

Comment:
Yeah, this is idiomatic. `other.borrow()` borrows; it doesn't consume. Basically, I don't *need* to take ownership of `StyleBasicShape` here, so I'm implementing `From<&StyleBasicShape>`. In such a case it's better to use `From<T: Borrow<StyleBasicShape>>` so that we don't need an `&StyleBasicShape` to call this (this leads to code like `(&foo).into()`, anything which can be converted to `&StyleBasicShape` works.